WebKit processes should not suspend during launch and initialization

pvollan opened the pull request:
#### 233820692ce5b0785c5631f10beb51b7fa9d99be
<pre>
WebKit processes should not suspend during launch and initialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=277163">https://bugs.webkit.org/show_bug.cgi?id=277163</a>
<a href="https://rdar.apple.com/132590940">rdar://132590940</a>

Reviewed by Chris Dumez.

Currently, there is a chance that WebKit processes on iOS will suspend between launch and initialization.
This is becaues we release the launch grant in AuxiliaryProcessProxy::didFinishLaunching, but we do not
take a foreground activity until later. For the WebContent process we do take a foreground activity when
initializing the process, but there is a chance the WebContent process can be suspended between the launch
and initialization events. For the Networking and GPU process, we do not take a foreground activity during
initialization. This is addressed by capturing an activity and launch grant when sending the message that
initializes the process, and releasing it after process initialization is finished. This is done for all
process types.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::initializeGPUProcess):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/ModelProcess/ModelProcess.messages.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::sendMessage):
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::didFinishProcessInitialization):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::ModelProcessProxy):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::initializeWebProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/281708@main">https://commits.webkit.org/281708@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/082d1adc4879216a860b090344ba3fec0150cdb5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60524 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64451 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11067 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48961 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7692 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62557 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37146 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52425 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29793 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33837 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9980 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55708 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9958 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66181 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9768 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56328 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4484 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56499 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3701 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9134 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35687 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36769 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37860 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36514 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -259,7 +259,7 @@ bool AuxiliaryProcessProxy::sendMessage(UniqueRef<IPC::Encoder>&& encoder, Optio
     if (asyncReplyHandler && canSendMessage() && shouldStartProcessThrottlerActivity == ShouldStartProcessThrottlerActivity::Yes) {
         auto completionHandler = WTFMove(asyncReplyHandler->completionHandler);
-        asyncReplyHandler->completionHandler = [activity = throttler().quietBackgroundActivity(description(encoder->messageName())), completionHandler = WTFMove(completionHandler)](IPC::Decoder* decoder) mutable {
+        asyncReplyHandler->completionHandler = [activity = throttler().foregroundActivity(description(encoder->messageName())), completionHandler = WTFMove(completionHandler)](IPC::Decoder* decoder) mutable {
```
2 issues here:
1. I don't think it is OK to take a foreground activity unconditionally. It should be on a case by case basis. If I send an async IPC to a background tab, e.g. to update some settings, there is no reason to give that background process *foreground* priority.
2. You want from a quiet activity to a non-quiet one, which means a lot of extra logging, which is bad for perf and power.

pvollan replied:
That's a very good point, Chris! I will remove the changes related to taking foreground activities during message sending and page loading, and only address the process suspension issue on launch in this PR.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -335,6 +335,8 @@ void AuxiliaryProcessProxy::didFinishLaunching(ProcessLauncher* launcher, IPC::C
     if (!connectionIdentifier)
         return;
 
+    m_initializationActivity = throttler().foregroundActivity("Process initialization Activity"_s).moveToUniquePtr();
```
Why take a foreground activity in didFinishLaunching, why not take it when starting to launch the process? Our ProcessAssertion logic will take care of taking the right assertion as soon as we get a PID already. There is no reason to delay the taking of the activity until we have a PID.

pvollan replied:
Good catch! I'll move this to where we start launching the process.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -277,6 +278,19 @@ class AuxiliaryProcessProxy
 
+    virtual void didFinishProcessInitialization()
+    {
+        WebCore::Timer::schedule(1_s, [weakThis = WeakPtr { this }]() {
```
We're not supposed to be using WebCore::Timer in the UIProcess because the app may mix WebKit1 and WebKit2. RunLoop::Timer is the one that is safe to use in the UIProcess.

cdumez commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -1541,7 +1541,7 @@ static _WKSOAuthorizationLoadPolicy wkSOAuthorizationLoadPolicy(SOAuthorizationL
         if (!m_networkActivity) {
             RELEASE_LOG(ProcessSuspension, "%p - NavigationState is taking a process network assertion because a page load started", this);
-            m_networkActivity = webView->_page->legacyMainFrameProcess().throttler().backgroundActivity("Page Load"_s).moveToUniquePtr();
+            m_networkActivity = webView->_page->legacyMainFrameProcess().throttler().foregroundActivity("Page Load"_s).moveToUniquePtr();
```
This is controversial I think. If you start a load in a visible WKWebView and background the app, do you really still expect the WebProcess to have a *foreground* assertion for up to 30 seconds? This has power implications.

pvollan replied:
Good point! I'll remove this change from the patch.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -602,7 +602,9 @@ static void logProcessPoolState(const WebProcessPool& pool)
-    pool->sendToAllProcesses(Messages::WebProcess::BacklightLevelDidChange(BKSDisplayBrightnessGetCurrent()));
+    float displayBrightness = BKSDisplayBrightnessGetCurrent();
+    pool->sendToAllProcesses(Messages::WebProcess::BacklightLevelDidChange(displayBrightness));
+    pool->m_displayBrightness = displayBrightness;
```
Not clear to me what this change is about.

pvollan replied:
This is caching the display brightness in the UI process, so we do not have to get this from the relevant service on each Web process launch. I will create a separate PR for this change.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -883,6 +883,9 @@ class WebProcessPool final
     RefPtr<StorageAccessPromptQuirkObserver> m_storageAccessPromptQuirksDataUpdateObserver;
 #endif
+#if PLATFORM(IOS_FAMILY) && !PLATFORM(MACCATALYST)
+    float m_displayBrightness { 0 };
```
Unclear how this relates to this patch exactly.

pvollan replied:
It is a very good point to break this up in several patches. I will address the concerns in follow-up patches.

Thanks for reviewing, Chris!

cdumez reviewed (changes requested):
This patch contains issues and controversial changes. It also tries to apply various optimization ideas all at once. It think we should choose a more fine-grained and careful approach, so we can quality each individual change.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -277,6 +277,15 @@ class AuxiliaryProcessProxy
     static RefPtr<WebCore::SharedBuffer> fetchAudioComponentServerRegistrations();
 #endif
 
+    virtual void didFinishProcessInitialization()
```
Why is this virtual? I don't see an override

pvollan replied:
Good point! I'll make it non-virtual in the next patch.

cdumez reviewed (changes requested):
This patch fails to update the ModelProcessProxy. This means that the initialization activity gets leaked. This is also evidence that it is error-prone and we should look into factoring the code differently so that such as mistake couldn't be made.

cdumez commented on Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
```diff
@@ -141,6 +141,7 @@ class ProcessLauncher : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<P
     const std::optional<ExtensionProcess>& extensionProcess() const { return m_process; }
     void setIsRetryingLaunch() { m_isRetryingLaunch = true; }
     bool isRetryingLaunch() const { return m_isRetryingLaunch; }
+    RefPtr<LaunchGrant> launchGrant() const { return m_launchGrant; }
```
nit: Should return a `LaunchGrant*` to avoid ref counting churn.

pvollan replied:
Ah, good point, will fix!

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -607,4 +607,14 @@ void AuxiliaryProcessProxy::didChangeThrottleState(ProcessThrottleState state)
 
+InitializationActivityAndGrant AuxiliaryProcessProxy::initializationActivityAndGrant() const
+{
+    return InitializationActivityAndGrant {
```
nit: Likely don't need `InitializationActivityAndGrant`

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -212,6 +212,7 @@ class AuxiliaryProcessProxy
 
 #if USE(EXTENSIONKIT)
     std::optional<ExtensionProcess> extensionProcess() const;
+    RefPtr<LaunchGrant> launchGrant() const;
```
Should return a `LaunchGrant*`

cdumez commented on Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
```diff
@@ -87,6 +87,13 @@
+{
+    if (!m_processLauncher)
+        return nullptr;
+    return m_processLauncher->launchGrant();
```
nit: could use a ternary.
```
return m_processLauncher ? m_processLauncher->launchGrant().get() : nullptr;
```

cdumez commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -212,7 +212,7 @@ GPUProcessProxy::GPUProcessProxy()
 
     // Initialize the GPU process.
-    send(Messages::GPUProcess::InitializeGPUProcess(WTFMove(parameters)), 0);
+    sendWithAsyncReply(Messages::GPUProcess::InitializeGPUProcess(WTFMove(parameters)), [weakThis = WeakPtr { *this }, initializationActivityAndGrant = initializationActivityAndGrant()] { });
```
Why are you capturing weakThis? Doesn't seem necessary.

cdumez commented on Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
```diff
@@ -87,7 +87,7 @@ ModelProcessProxy::ModelProcessProxy()
 
     // Initialize the model process.
-    send(Messages::ModelProcess::InitializeModelProcess(WTFMove(parameters)), 0);
+    sendWithAsyncReply(Messages::ModelProcess::InitializeModelProcess(WTFMove(parameters)), [weakThis = WeakPtr { *this }, initializationActivityAndGrant = initializationActivityAndGrant()] () { }, 0);
```
Why are you capturing weakThis here?

pvollan replied:
Good point! This is not needed.

cdumez reviewed (approved):
LGTM with nits.

pvollan committed:
WebKit processes should not suspend during launch and initialization
https://bugs.webkit.org/show_bug.cgi?id=277163
rdar://132590940

Reviewed by Chris Dumez.

Currently, there is a chance that WebKit processes on iOS will suspend between launch and initialization.
This is becaues we release the launch grant in AuxiliaryProcessProxy::didFinishLaunching, but we do not
take a foreground activity until later. For the WebContent process we do take a foreground activity when
initializing the process, but there is a chance the WebContent process can be suspended between the launch
and initialization events. For the Networking and GPU process, we do not take a foreground activity during
initialization. This is addressed by capturing an activity and launch grant when sending the message that
initializes the process, and releasing it after process initialization is finished. This is done for all
process types.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::initializeGPUProcess):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/ModelProcess/ModelProcess.messages.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::sendMessage):
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::didFinishProcessInitialization):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::ModelProcessProxy):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::initializeWebProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: https://commits.webkit.org/281708@main
